src/vcpkg/portfileprovider.cpp

```diff
@@ -10,10 +11,288 @@
+                    if (already_in_out == out.end() || already_in_out->first != maybe_this_port.first)
+                    {
+                        out.emplace_hint(already_in_out, maybe_this_port.first, this_port);
+                    }
+
+                    return Unit{};
+                }
+
+                return maybe_this_port.second.error();
+            }
+            case OverlayPortKind::Directory:
+            {
+                auto maybe_subdirectories = fs.try_get_directories_non_recursive(m_directory);
+                if (auto subdirectories = maybe_subdirectories.get())
+                {
+                    std::vector<LocalizedString> errors;
+                    Util::sort(*subdirectories);
+                    auto first_loaded = m_loaded_ports.begin();
+                    const auto last_loaded = m_loaded_ports.end();
+                    auto first_out = out.begin();
+                    const auto last_out = out.end();
+                    for (const auto& full_subdirectory : *subdirectories)
+                    {
+                        auto subdirectory = full_subdirectory.filename();
+                        while (first_out != last_out && first_out->first < subdirectory)
+                        {
+                            ++first_out;
+                        }
+
+                        if (first_out != last_out && first_out->first == subdirectory)
+                        {
+                            // this subdirectory is already in the output; we shouldn't replace or attempt to load it
+                            ++first_out;
+                            continue;
+                        }
+
+                        while (first_loaded != last_loaded && first_loaded->first < subdirectory)
+                        {
+                            ++first_loaded;
+                        }
+
+                        if (first_loaded == last_loaded || first_loaded->first != subdirectory)
+                        {
+                            // the subdirectory isn't cached, load it into the cache
+                            first_loaded = try_load_port_subdirectory_uncached(first_loaded, fs, subdirectory);
+                        } // else: the subdirectory is already loaded
+
+                        if (auto this_port = first_loaded->second.get())
+                        {
+                            first_out = out.emplace_hint(first_out, first_loaded->first, this_port);
+                            ++first_out;
+                        }
+                        else
+                        {
+                            errors.push_back(first_loaded->second.error());
+                        }
+
+                        ++first_loaded;
+                    }
+
+                    if (errors.empty())
+                    {
+                        return Unit{};
+                    }
+
+                    return LocalizedString::from_raw(Strings::join("\n", errors));
```

Comment:
I think stopping at the first error rather than trying to return all of them would also be reasonable behavior, but that broke the `ci` e2e test so I kept this join behavior. (I did not keep it for different values of `--overlay-ports`....)